src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscriptionsConfigExtensions.cs

```diff
@@ -82,15 +91,26 @@ public static class MessageDrivenSubscriptionsConfigExtensions
         public static void RegisterPublisherForAssembly<T>(this RoutingSettings<T> routingSettings, Assembly eventAssembly, string eventNamespace, string publisherEndpoint) where T : TransportDefinition, IMessageDrivenSubscriptionTransport
         {
             Guard.AgainstNull(nameof(eventAssembly), eventAssembly);
-            Guard.AgainstNull(nameof(eventNamespace), eventNamespace);
             Guard.AgainstNullAndEmpty(nameof(publisherEndpoint), publisherEndpoint);
 
             ThrowOnAddress(publisherEndpoint);
 
-            routingSettings.Settings.GetOrCreate<Publishers>().Add(eventAssembly, eventNamespace, publisherEndpoint);
-        }
+            // empty namespace is null, not string.empty
+            eventNamespace = eventNamespace == string.Empty ? null : eventNamespace;
 
-        static void ThrowOnAddress(string publisherEndpoint)
+            routingSettings.Settings.GetOrCreate<ConfiguredPublishers>().Add((publishers, knownMessageTypes) =>
+            {
+                foreach (var knownMessageType in knownMessageTypes)
+                {
+                    if (knownMessageType.Assembly == eventAssembly && knownMessageType.Namespace == eventNamespace)
+                    {
+                        publishers.Add(knownMessageType, publisherEndpoint);
```

Comment:
@danielmarbach I think a shared code here would complicate things mainly because we add the `Publishers` or the `UnicastRoutingTable` which are two entirely different classes (which is something we should maybe review, but out of scope). I'd vote to go with the current proposal